src/votekit/ballot_graph.py

```diff
@@ -0,0 +1,226 @@
+            if n == 3:
+                Gc.add_edges_from([(k,tuple([i])) for k in G_corner.nodes])
+            else:
+                Gc.add_edges_from([(k,tuple([i])) for k in G_corner.nodes if len(k) == 2])
+            
+        nodes = Gc.nodes
+
+            # add the additional edges corresponding to swapping the order of the
+            # first two candidates
+        new_edges = []
+        for k in nodes:
+            if len(k)==2:
+                new_edges.append(((k[0],k[1]),(k[1],k[0])))
+            elif len(k)>2:
+                l = list(k)
+                a = l[0]
+                b=l[1]
+                new_edges.append((tuple([a]+[b]+l[2:]), tuple([b]+[a]+l[2:])))
+
+        Gc.add_edges_from(new_edges)
+
+    Graphs[n] = Gc
+    return Gc
+
+
+def relabel(gr, new_label, num_cands):
+    node_map = {}
+    graph_nodes = list(gr.nodes)
+
+
+    for k in graph_nodes:
+            # add the value of new_label to every entry in every ballot
+        tmp = [new_label+y for y in k]
+
+            # reduce everything mod new_label
+        for i in range(len(tmp)):
+            if tmp[i]> num_cands:
+                tmp[i] = tmp[i]- num_cands
+        node_map[k] = tuple([new_label]+tmp)
+
+    return nx.relabel_nodes(gr, node_map)
+
+
+Graphs = {}
+
+
+class BallotGraph():
+    profile: PreferenceProfile
+    ballot_dict: dict
+    num_cands: int
+    num_voters: int
+    
+    
+    def __init__(self,profile):
+        self.profile = profile
+        self.ballot_dict= profile.to_dict()
+        
+        #cand_list = []
+        #for key in ballot_dict.keys():
+        #    for i in key:
+        #        cand_list.append(i)
+      
+        self.num_cands = len(self.profile.get_candidates())
+        
+        if self.num_cands not in Graphs.keys():
+            build_graph(self.num_cands)
+        
+        
+        all_ballots = Graphs[self.num_cands].nodes
+        di = {}
+        for ballot in all_ballots:
+            di[ballot] = 0
+        
+        self.ballot_dict = di | ballot_dict
+        
+        
+        self.clean()
+        
+        self.num_voters = sum(ballot_dict.values())
+        
+        
+    def clean(self): #deletes empty ballots, changes n-1 length ballots to n length ballots and updates counts
+        di = self.ballot_dict.copy()
+        
+        for ballot in di.keys():
+            if len(ballot)==0:
+                self.ballot_dict.pop(ballot)
+            elif len(ballot)==self.num_cands-1:
+                for i in self.profile.get_candidates():
+                    if i not in ballot:
+                        self.ballot_dict[ballot+tuple([i])]+= di[ballot]    
+                        self.ballot_dict.pop(ballot)
+                        break
+                
+        
+    def visualize(self, neighborhoods = {}):
+        Gc = Graphs[self.num_cands]    
+        if neighborhoods == {}:
+            self.clean()
+            WHITE = (1,1,1)
```

Comment:
Will these WHITE/BLACK variables be the same for every instance a visualized graph or is that something the user should be able to input?

Reply:
no they're gonna be the same for every instance. They just specify the colors that you don't want to use when visualizing. White doesn't make sense, and using black would mess with the visibility of the node labels